src/OpenTracing/Mock/MockTracer.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Collections.Generic;
+using OpenTracing.Noop;
+using OpenTracing.Propagation;
+using OpenTracing.Util;
+
+namespace OpenTracing.Mock
+{
+    /// <summary>
+    /// <see cref="MockTracer"/> makes it easy to test the semantics of OpenTracing instrumentation.
+    /// <para/>
+    /// By using a <see cref="MockTracer"/> as an <see cref="ITracer"/> implementation for unittests, a developer can assert that Span
+    /// properties and relationships with other Spans are defined as expected by instrumentation code.
+    /// <para/>
+    /// The MockTracerTests class has simple usage examples.
+    /// </summary>
+    public class MockTracer : ITracer
+    {
+        private readonly object _lock = new object();
+
+        private readonly List<MockSpan> _finishedSpans = new List<MockSpan>();
+        private readonly IPropagator _propagator;
+        private readonly IScopeManager _scopeManager;
+
+        public IScopeManager ScopeManager => _scopeManager;
+
+        public ISpan ActiveSpan => _scopeManager?.Active?.Span;
+
+        public MockTracer()
+            : this(new AsyncLocalScopeManager(), Propagators.TextMap)
+        {
+        }
+
+        public MockTracer(IScopeManager scopeManager)
+            : this(scopeManager, Propagators.TextMap)
+        {
+        }
+
+        public MockTracer(IScopeManager scopeManager, IPropagator propagator)
+        {
+            _scopeManager = scopeManager;
+            _propagator = propagator;
+        }
+
+        /// <summary>
+        /// Create a new <see cref="MockTracer"/> that passes through any calls
+        /// to <see cref="ITracer.Inject"/> and/or <see cref="ITracer.Extract"/>.
+        /// </summary>
+        public MockTracer(IPropagator propagator)
+            : this(NoopScopeManager.Instance, propagator)
```

Comment:
Should we default to AsyncLocalScopeManager here to match the parameterless constructor?

Reply:
+1 on this. The one from Java effectively uses `ThreadLocalScopeManager`, which is the default/defacto manager, and `AsyncLocalScopeManager` is the equivalent one.

Reply:
For this overload, Java [uses the `NoopScopeManager` as well](https://github.com/opentracing/opentracing-java/blob/master/opentracing-mock/src/main/java/io/opentracing/mock/MockTracer.java#L62). Changing this would be a functional difference so I'd prefer not to do that for now.